DigitalLearningSolutions.Web/ServiceFilter/ValidateAllowedDlsSubApplication.cs

```diff
@@ -0,0 +1,134 @@
+﻿namespace DigitalLearningSolutions.Web.ServiceFilter
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using DigitalLearningSolutions.Data.Enums;
+    using DigitalLearningSolutions.Web.Controllers;
+    using DigitalLearningSolutions.Web.Controllers.LearningSolutions;
+    using DigitalLearningSolutions.Web.Helpers;
+    using DigitalLearningSolutions.Web.Models.Enums;
+    using Microsoft.AspNetCore.Mvc;
+    using Microsoft.AspNetCore.Mvc.Filters;
+    using Microsoft.AspNetCore.Mvc.ModelBinding;
+    using Microsoft.FeatureManagement;
+
+    public class ValidateAllowedDlsSubApplication : IActionFilter
+    {
+        private readonly IFeatureManager featureManager;
+        private readonly IEnumerable<DlsSubApplication> validApplications;
+
+        public ValidateAllowedDlsSubApplication(
+            string[] validApplicationNames,
+            IFeatureManager featureManager
+        )
+        {
+            validApplications = validApplicationNames.Select(Enumeration.FromName<DlsSubApplication>);
+            this.featureManager = featureManager;
+        }
+
+        public void OnActionExecuted(ActionExecutedContext context) { }
+
+        public async void OnActionExecuting(ActionExecutingContext context)

```

Comment:
It would be nice if we could pull a couple of methods out of this.
I think some stages which could possibly be pulled out are
* Redirect user if application param not recognised
* Parse application param into DlsSubApplication
* Validate whether this page exists on this application
* Validate whether user has permission to access this application

Can we make some improvements in this direction?

Reply:
I've reorganised it into something that looks clearer, but I'm not sure whether two out variables is best practice, my intuition dislikes them.

Reply:
The top level definitely looks nicer, but I think you're right, we can improve that method with two out variables.